File: argus/packs/registry.py
"""Install, list and remove knowledge packs.

There is no registry file. A pack is already self-describing, so the installed
files *are* the registry: "registered" and "present and readable" are the same
statement, and there is no index that can drift out of step with the directory
it describes.

Installation is verify-then-place. A pack is streamed to a temporary file in
the destination directory, checked, and only then renamed into position, so a
failed install leaves nothing behind -- a truncated download that quietly
became a half-empty knowledge base is the exact "looks like it works" failure
this format is built to avoid.
"""

from __future__ import annotations

import hashlib
import json
import re
import uuid
from dataclasses import dataclass
from pathlib import Path

import httpx

from .. import embed as embed_module
from . import format as pack_format

PACK_SUFFIX = ".arguspack"

#: Pack names become filenames, and a pack's name comes from pack_meta inside
#: a file that may have been downloaded from anyone. Without this an installed
#: pack could name itself "../../.ssh/authorized_keys" and be written straight
#: out of the destination directory.
_NAME_RE = re.compile(r"^[a-z0-9][a-z0-9._-]{0,63}$")

_DOWNLOAD_CHUNK = 1 << 20
_INDEX_TIMEOUT = 30.0
_DOWNLOAD_TIMEOUT = 300.0


class RegistryError(Exception):
    """A pack could not be installed, listed or removed."""


@dataclass(frozen=True)
class InstalledPack:
    name: str
    version: str
    path: Path
    embedding_model: str
    embedding_dim: str
    size_bytes: int
    license: str
    attribution: str
    source_commit: str
    #: False when the pack's embedding space differs from this instance's.
    #: Lexical and symbol search still work; only semantic search is affected.
    compatible: bool
    incompatible_reason: str = ""


@dataclass(frozen=True)
class IndexEntry:
    name: str
    version: str
    url: str
    sha256: str
    size_bytes: int
    license: str


def install(
    url_or_path: str | Path, *, dest_dir: Path, expected_sha256: str | None = None,
    client: httpx.Client | None = None,
) -> InstalledPack:
    """Fetch or copy a pack into ``dest_dir`` after verifying it."""
    dest_dir = Path(dest_dir)
    dest_dir.mkdir(parents=True, exist_ok=True)

    # Staged inside dest_dir so the final move is a rename, not a copy across
    # filesystems -- a copy could be observed half-done.
    staging = dest_dir / f".incoming-{uuid.uuid4().hex}.tmp"

    try:
        digest = _stage(url_or_path, staging, client=client)

        if expected_sha256 and digest.lower() != expected_sha256.strip().lower():
            raise RegistryError(
                f"checksum mismatch for {url_or_path}: expected "
                f"{expected_sha256.strip().lower()}, got {digest}"
            )

        meta = _read_pack_meta(staging)
        name = _require_name(meta.get("source_name", ""))
        compatible, reason = _compatibility(meta)

        final = dest_dir / f"{name}{PACK_SUFFIX}"
        staging.replace(final)
    except BaseException:
        staging.unlink(missing_ok=True)
        raise

    return InstalledPack(
        name=name,
        version=meta.get("pack_version", ""),
        path=final,
        embedding_model=meta.get("embedding_model", ""),
        embedding_dim=meta.get("embedding_dim", ""),
        size_bytes=final.stat().st_size,
        license=meta.get("license", ""),
        attribution=meta.get("attribution", ""),
        source_commit=meta.get("source_commit", ""),
        compatible=compatible,
        incompatible_reason=reason,
    )


def list_installed(dest_dir: Path) -> list[InstalledPack]:
    """Every pack in ``dest_dir``, including any that cannot be read.

    An unreadable file is reported as incompatible rather than skipped. Silently
    omitting it would present a knowledge base that is quietly missing a source,
    which is the failure this module exists to prevent.
    """
    dest_dir = Path(dest_dir)
    if not dest_dir.is_dir():
        return []

    packs = []
    for path in sorted(dest_dir.glob(f"*{PACK_SUFFIX}")):
        try:
            meta = _read_pack_meta(path)
        except RegistryError as exc:
            packs.append(InstalledPack(
                name=path.stem, version="", path=path,
                embedding_model="", embedding_dim="",
                size_bytes=path.stat().st_size, license="", attribution="",
                source_commit="", compatible=False,
                incompatible_reason=f"unreadable: {exc}",
            ))
            continue
        compatible, reason = _compatibility(meta)
        packs.append(InstalledPack(
            name=meta.get("source_name", path.stem),
            version=meta.get("pack_version", ""),
            path=path,
            embedding_model=meta.get("embedding_model", ""),
            embedding_dim=meta.get("embedding_dim", ""),
            size_bytes=path.stat().st_size,
            license=meta.get("license", ""),
            attribution=meta.get("attribution", ""),
            source_commit=meta.get("source_commit", ""),
            compatible=compatible,
            incompatible_reason=reason,
        ))
    return packs


def remove(name: str, dest_dir: Path) -> bool:
    """Delete an installed pack. True if one was removed, False if absent."""
    path = Path(dest_dir) / f"{_require_name(name)}{PACK_SUFFIX}"
    if not path.is_file():
        return False
    path.unlink()
    return True


def fetch_index(url: str, *, client: httpx.Client | None = None) -> list[IndexEntry]:
    """Read a published index of available packs."""
    owns_client = client is None
    client = client or httpx.Client(timeout=_INDEX_TIMEOUT)
    try:
        response = client.get(url)
    except httpx.HTTPError as exc:
        raise RegistryError(f"GET {url} failed: {exc}") from exc
    finally:
        if owns_client:
            client.close()

    if response.status_code != 200:
        raise RegistryError(
            f"GET {url} returned {response.status_code}: {response.text[:200]}"
        )
    try:
        body = response.json()
    except json.JSONDecodeError as exc:
        raise RegistryError(
            f"GET {url}: index is not JSON: {response.text[:200]}"
        ) from exc

    raw = body.get("packs") if isinstance(body, dict) else None
    if not isinstance(raw, list):
        raise RegistryError(f"GET {url}: index has no 'packs' list")

    entries = []
    for item in raw:
        if not isinstance(item, dict):
            raise RegistryError(f"GET {url}: index entry is not an object: {item!r}")
        missing = [k for k in ("name", "version", "url", "sha256") if not item.get(k)]
        if missing:
            # A published entry without a checksum cannot be verified on
            # download, so it is a malformed index rather than an optional
            # field.
            raise RegistryError(
                f"GET {url}: index entry {item.get('name', '?')!r} is missing "
                f"{', '.join(missing)}"
            )
        entries.append(IndexEntry(
            name=str(item["name"]),
            version=str(item["version"]),
            url=str(item["url"]),
            sha256=str(item["sha256"]),
            size_bytes=int(item.get("size_bytes") or 0),
            license=str(item.get("license", "")),
        ))
    return entries


def _stage(
    url_or_path: str | Path, staging: Path, *, client: httpx.Client | None,
) -> str:
    """Write the pack to ``staging``; return its hex SHA-256."""
    text = str(url_or_path)
    if text.startswith(("http://", "https://")):
        return _download(text, staging, client=client)

    source = Path(url_or_path)
    if not source.is_file():
        raise RegistryError(f"no such pack file: {source}")
    with source.open("rb") as src, staging.open("wb") as dst:
        return _copy_hashing(src, dst)


def _download(url: str, staging: Path, *, client: httpx.Client | None) -> str:
    owns_client = client is None
    client = client or httpx.Client(timeout=_DOWNLOAD_TIMEOUT, follow_redirects=True)
    try:
        with client.stream("GET", url) as response:
            if response.status_code != 200:
                raise RegistryError(f"GET {url} returned {response.status_code}")
            digest = hashlib.sha256()
            with staging.open("wb") as handle:
                for chunk in response.iter_bytes(_DOWNLOAD_CHUNK):
                    digest.update(chunk)
                    handle.write(chunk)
            return digest.hexdigest()
    except httpx.HTTPError as exc:
        raise RegistryError(f"GET {url} failed: {exc}") from exc
    finally:
        if owns_client:
            client.close()


def _copy_hashing(src, dst) -> str:
    digest = hashlib.sha256()
    while True:
        chunk = src.read(_DOWNLOAD_CHUNK)
        if not chunk:
            return digest.hexdigest()
        digest.update(chunk)
        dst.write(chunk)


def _read_pack_meta(path: Path) -> dict[str, str]:
    try:
        conn = pack_format.open_pack(path)
    except Exception as exc:
        raise RegistryError(f"{path.name} is not a readable pack: {exc}") from exc
    try:
        meta = pack_format.read_meta(conn)
    except Exception as exc:
        raise RegistryError(f"{path.name} has no readable pack_meta: {exc}") from exc
    finally:
        conn.close()
    if not meta:
        raise RegistryError(f"{path.name} has an empty pack_meta")
    return meta


def _require_name(name: str) -> str:
    if not _NAME_RE.match(name or ""):
        raise RegistryError(
            f"invalid pack name {name!r}: names must match {_NAME_RE.pattern} "
            f"(a pack names its own file, so this is what keeps a downloaded "
            f"pack inside the directory it was installed into)"
        )
    return name


def _compatibility(meta: dict[str, str]) -> tuple[bool, str]:
    try:
        pack_format.require_compatible(
            meta, model=embed_module.EMBED_MODEL, dim=embed_module.EMBED_DIM
        )
    except pack_format.PackMismatch as exc:
        return False, str(exc)
    return True, ""

File: src/Argus/Packs/Registry.cs
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using Argus.Indexing;
using Argus.Util;

namespace Argus.Packs;

/// <summary>A pack could not be installed, listed, fetched or indexed.</summary>
public sealed class RegistryError(string message, Exception? inner = null) : Exception(message, inner);

public sealed record InstalledPack(
    string Name, string Version, string Path, string EmbeddingModel, string EmbeddingDim, long SizeBytes,
    string License, string Attribution, string SourceCommit, bool Compatible, string IncompatibleReason = "");

public sealed record IndexEntry(string Name, string Version, string Url, string Sha256, long SizeBytes, string License);

/// <summary>
/// Installing, listing and publishing packs. A pack
/// names its own file from its metadata, so the name is validated before it is
/// used as a path; an install stages to a temp file and renames into place, so a
/// failed download never replaces a working pack.
/// </summary>
public static class Registry
{
    public const string PackSuffix = ".arguspack";
    static readonly Regex NameRe = new(@"^[a-z0-9][a-z0-9._-]{0,63}$", RegexOptions.CultureInvariant);
    const int DownloadChunk = 1 << 20;

    /// <summary>Tests replace this to serve downloads from memory.</summary>
    public static Func<HttpMessageHandler>? HandlerOverride { get; set; }

    static HttpClient Client(double timeoutSeconds) =>
        new(HandlerOverride?.Invoke() ?? new HttpClientHandler { AllowAutoRedirect = true }, disposeHandler: true)
            { Timeout = TimeSpan.FromSeconds(timeoutSeconds) };

    public static InstalledPack Install(string urlOrPath, string destDir, string? expectedSha256 = null)
    {
        Directory.CreateDirectory(destDir);
        var staging = Path.Combine(destDir, $".incoming-{Guid.NewGuid():N}.tmp");
        string name, finalPath;
        Dictionary<string, string> meta;
        bool compatible;
        string reason;
        try
        {
            var digest = Stage(urlOrPath, staging);
            if (!string.IsNullOrEmpty(expectedSha256) && digest.ToLowerInvariant() != expectedSha256.Trim().ToLowerInvariant())
                throw new RegistryError($"checksum mismatch for {urlOrPath}: expected {expectedSha256.Trim().ToLowerInvariant()}, got {digest}");
            meta = ReadPackMeta(staging);
            name = RequireName(meta.GetValueOrDefault("source_name", ""));
            (compatible, reason) = Compatibility(meta);
            finalPath = Path.Combine(destDir, $"{name}{PackSuffix}");
            File.Move(staging, finalPath, overwrite: true);
        }
        catch
        {
            try { File.Delete(staging); } catch (IOException) { }
            throw;
        }
        return new InstalledPack(name, meta.GetValueOrDefault("pack_version", ""), finalPath,
            meta.GetValueOrDefault("embedding_model", ""), meta.GetValueOrDefault("embedding_dim", ""),
            new FileInfo(finalPath).Length, meta.GetValueOrDefault("license", ""), meta.GetValueOrDefault("attribution", ""),
            meta.GetValueOrDefault("source_commit", ""), compatible, reason);
    }

    public static List<string> PackFiles(string dir) =>
        Directory.Exists(dir)
            ? Directory.EnumerateFiles(dir, "*" + PackSuffix).OrderBy(p => p, StringComparer.Ordinal).ToList()
            : [];

    public static List<InstalledPack> ListInstalled(string destDir)
    {
        var packs = new List<InstalledPack>();
        foreach (var path in PackFiles(destDir))
        {
            var stem = Path.GetFileNameWithoutExtension(path);
            Dictionary<string, string> meta;
            try { meta = ReadPackMeta(path); }
            catch (RegistryError exc)
            {
                packs.Add(new InstalledPack(stem, "", path, "", "", new FileInfo(path).Length, "", "", "", false, $"unreadable: {exc.Message}"));
                continue;
            }
            var (compatible, reason) = Compatibility(meta);
            packs.Add(new InstalledPack(meta.GetValueOrDefault("source_name", stem), meta.GetValueOrDefault("pack_version", ""), path,
                meta.GetValueOrDefault("embedding_model", ""), meta.GetValueOrDefault("embedding_dim", ""), new FileInfo(path).Length,
                meta.GetValueOrDefault("license", ""), meta.GetValueOrDefault("attribution", ""), meta.GetValueOrDefault("source_commit", ""),
                compatible, reason));
        }
        return packs;
    }

    public static bool Remove(string name, string destDir)
    {
        var path = Path.Combine(destDir, $"{RequireName(name)}{PackSuffix}");
        if (!File.Exists(path)) return false;
        File.Delete(path);
        return true;
    }

    public static List<IndexEntry> FetchIndex(string url)
    {
        HttpResult response;
        try
        {
            using var client = Client(30.0);
            using var resp = client.GetAsync(url).GetAwaiter().GetResult();
            response = new HttpResult((int)resp.StatusCode, resp.Content.ReadAsStringAsync().GetAwaiter().GetResult());
        }
        catch (Exception exc) when (exc is HttpRequestException or TaskCanceledException or InvalidOperationException or UriFormatException)
        {
            throw new RegistryError($"GET {url} failed: {exc.Message}", exc);
        }
        if (response.Status != 200) throw new RegistryError($"GET {url} returned {response.Status}: {PyStr.Prefix(response.Text, 200)}");
        JsonNode? body;
        try { body = JsonNode.Parse(response.Text); }
        catch (JsonException exc) { throw new RegistryError($"GET {url}: index is not JSON: {PyStr.Prefix(response.Text, 200)}", exc); }
        if ((body as JsonObject)?["packs"] is not JsonArray raw) throw new RegistryError($"GET {url}: index has no 'packs' list");
        var entries = new List<IndexEntry>();
        foreach (var item in raw)
        {
            if (item is not JsonObject obj) throw new RegistryError($"GET {url}: index entry is not an object: {item?.ToJsonString()}");
            var missing = new[] { "name", "version", "url", "sha256" }.Where(k => !Truthy(obj[k])).ToList();
            if (missing.Count > 0)
                throw new RegistryError($"GET {url}: index entry {PyStr.Repr(obj["name"]?.ToString() ?? "?")} is missing {string.Join(", ", missing)}");
            entries.Add(new IndexEntry(obj["name"]!.ToString(), obj["version"]!.ToString(), obj["url"]!.ToString(), obj["sha256"]!.ToString(),
                Truthy(obj["size_bytes"]) ? Convert.ToInt64(obj["size_bytes"]!.ToString()) : 0, obj["license"]?.ToString() ?? ""));
        }
        return entries;
    }

    static bool Truthy(JsonNode? n) => n switch
    {
        null => false,
        JsonValue v when v.TryGetValue<string>(out var s) => s.Length > 0,
        JsonValue v when v.TryGetValue<bool>(out var b) => b,
        JsonValue v when v.TryGetValue<double>(out var d) => d != 0,
        JsonArray a => a.Count > 0,
        JsonObject o => o.Count > 0,
        _ => true,
    };

    public static JsonObject WriteIndex(string destDir, string baseUrl, string? name = null)
    {
        var b = baseUrl.TrimEnd('/');
        var entries = new JsonArray();
        var skipped = new JsonArray();
        foreach (var path in PackFiles(destDir))
        {
            Dictionary<string, string> meta;
            try
            {
                using var conn = PackFormat.OpenPack(path);
                meta = PackFormat.ReadMeta(conn);
            }
            catch (Exception exc)
            {
                skipped.Add(new JsonObject { ["file"] = Path.GetFileName(path), ["reason"] = PyStr.Prefix($"{exc.GetType().Name}: {exc.Message}", 120) });
                continue;
            }
            var entryName = meta.GetValueOrDefault("source_name", "");
            if (name is not null && entryName != name) continue;
            if (entryName.Length == 0)
            {
                skipped.Add(new JsonObject { ["file"] = Path.GetFileName(path), ["reason"] = "pack records no source_name" });
                continue;
            }
            entries.Add(new JsonObject
            {
                ["name"] = entryName,
                ["version"] = meta.GetValueOrDefault("pack_version", ""),
                ["url"] = b.Length > 0 ? $"{b}/{Path.GetFileName(path)}" : Path.GetFileName(path),
                ["sha256"] = Sha256File(path),
                ["size_bytes"] = new FileInfo(path).Length,
                ["license"] = meta.GetValueOrDefault("license", ""),
                ["attribution"] = meta.GetValueOrDefault("attribution", ""),
                ["source_commit"] = meta.GetValueOrDefault("source_commit", ""),
            });
        }
        return new JsonObject { ["schema"] = 1, ["packs"] = entries, ["skipped"] = skipped };
    }

    public static string Sha256File(string path)
    {
        using var stream = File.OpenRead(path);
        return Convert.ToHexStringLower(SHA256.HashData(stream));
    }

    static string Stage(string urlOrPath, string staging)
    {
        if (urlOrPath.StartsWith("http://", StringComparison.Ordinal) || urlOrPath.StartsWith("https://", StringComparison.Ordinal))
            return Download(urlOrPath, staging);
        if (!File.Exists(urlOrPath)) throw new RegistryError($"no such pack file: {urlOrPath}");
        using var src = File.OpenRead(urlOrPath);
        using var dst = File.Create(staging);
        return CopyHashing(src, dst);
    }

    static string Download(string url, string staging)
    {
        try
        {
            using var client = Client(300.0);
            using var resp = client.GetAsync(url, HttpCompletionOption.ResponseHeadersRead).GetAwaiter().GetResult();
            if ((int)resp.StatusCode != 200) throw new RegistryError($"GET {url} returned {(int)resp.StatusCode}");
            using var src = resp.Content.ReadAsStream();
            using var dst = File.Create(staging);
            return CopyHashing(src, dst);
        }
        catch (Exception exc) when (exc is HttpRequestException or TaskCanceledException or IOException)
        {
            throw new RegistryError($"GET {url} failed: {exc.Message}", exc);
        }
    }

    static string CopyHashing(Stream src, Stream dst)
    {
        using var hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        var buffer = new byte[DownloadChunk];
        int n;
        while ((n = src.Read(buffer, 0, buffer.Length)) > 0)
        {
            hash.AppendData(buffer, 0, n);
            dst.Write(buffer, 0, n);
        }
        return Convert.ToHexStringLower(hash.GetHashAndReset());
    }

    static Dictionary<string, string> ReadPackMeta(string path)
    {
        Microsoft.Data.Sqlite.SqliteConnection conn;
        try { conn = PackFormat.OpenPack(path); }
        catch (Exception exc) { throw new RegistryError($"{Path.GetFileName(path)} is not a readable pack: {exc.Message}", exc); }
        Dictionary<string, string> meta;
        try { meta = PackFormat.ReadMeta(conn); }
        catch (Exception exc) { throw new RegistryError($"{Path.GetFileName(path)} has no readable pack_meta: {exc.Message}", exc); }
        finally { conn.Dispose(); }
        if (meta.Count == 0) throw new RegistryError($"{Path.GetFileName(path)} has an empty pack_meta");
        return meta;
    }

    static string RequireName(string name)
    {
        if (!NameRe.IsMatch(name ?? ""))
            throw new RegistryError(
                $"invalid pack name {PyStr.Repr(name)}: names must match {NameRe} (a pack names its own file, so this is what keeps a downloaded pack inside the directory it was installed into)");
        return name!;
    }

    static (bool, string) Compatibility(Dictionary<string, string> meta)
    {
        try { PackFormat.RequireCompatible(meta, Embed.Model, Embed.Dim); }
        catch (PackMismatch exc) { return (false, exc.Message); }
        return (true, "");
    }
}

/// <summary>
/// A persistent cache of chunk embeddings for pack builds.
/// Every failure degrades to "not cached": the cache is an accelerator, never a
/// dependency.
/// </summary>
public sealed class EmbeddingCache : IDisposable
{
    public const int CacheFormat = 1;
    Microsoft.Data.Sqlite.SqliteConnection? _conn;
    public int Hits { get; private set; }
    public int Misses { get; private set; }

    public static string CacheKey(string text, string model, int dim) =>
        $"{CacheFormat}:{model}:{dim}:{Convert.ToHexStringLower(SHA256.HashData(System.Text.Encoding.UTF8.GetBytes(text)))}";

    public EmbeddingCache(string? path)
    {
        if (path is null) return;
        try
        {
            Store.Db.Init();
            Directory.CreateDirectory(Path.GetDirectoryName(Path.GetFullPath(path))!);
            var conn = new Microsoft.Data.Sqlite.SqliteConnection(new Microsoft.Data.Sqlite.SqliteConnectionStringBuilder { DataSource = path, Pooling = false }.ToString());
            conn.Open();
            Sql.Script(conn, "PRAGMA journal_mode=WAL; CREATE TABLE IF NOT EXISTS embeddings (key TEXT PRIMARY KEY, bits BLOB NOT NULL, i8 BLOB NOT NULL);");
            _conn = conn;
        }
        catch (Exception exc) when (exc is Microsoft.Data.Sqlite.SqliteException or IOException or UnauthorizedAccessException)
        {
            _conn = null;
        }
    }

    public bool Enabled => _conn is not null;

    public (byte[] Bits, byte[] I8)? Get(string key)
    {
        if (_conn is null) return null;
        Row? row;
        try { row = Sql.One(_conn, "SELECT bits, i8 FROM embeddings WHERE key = ?", key); }
        catch (Microsoft.Data.Sqlite.SqliteException) { return null; }
        if (row is null) { Misses++; return null; }
        Hits++;
        return (row.Bytes("bits")!, row.Bytes("i8")!);
    }

    public void PutMany(IReadOnlyList<(string Key, byte[] Bits, byte[] I8)> rows)
    {
        if (_conn is null || rows.Count == 0) return;
        try
        {
            using var tx = _conn.BeginTransaction();
            foreach (var (k, b, i) in rows)
                Sql.Exec(_conn, "INSERT OR REPLACE INTO embeddings (key, bits, i8) VALUES (?, ?, ?)", k, b, i);
            tx.Commit();
        }
        catch (Microsoft.Data.Sqlite.SqliteException) { }
    }

    public void Dispose()
    {
        _conn?.Dispose();
        _conn = null;
    }
}
